### DISCUSSION TITLE
Deep recursion on subroutine "DB::_get_reference_descriptor

### DISCUSSION THREAD

**Discussion by markAtAthena** [Camelcade/Devel-Camelcadedb#56] [General]
I just ran into [2021-02-05 17:28:14 2456] [warn] Deep recursion on subroutine "DB::_get_reference_descriptor" at /home/mark/perl5/lib/perl5/Devel/Camelcadedb.pm line 584.

and then shortly thereafter

Out of memory!
Out of memory!
Out of memory!

which I suspect is covered by this:

https://github.com/Camelcade/Devel-Camelcadedb/issues/38

are there any workarounds?

**Comment by hurricup** [comment]
Sorry, but I could not investigate this yet :(

**Reply by markAtAthena** [reply]
#38 did get rid of the deep recursion, but now this happens instead as soon as I exercise my code in the same way that triggered the previous deep recursion:

The debugger disconnects and I see this in my remote ssh terminal where I started the perl process from:

Buffer  has no newlines in it and nothing is in the socket

**Reply by markAtAthena** [reply]
The perl process continues to run, but the result in the process is a reported error (that doesn't occur if I do not try to debug it and do the same actions) and I can't reconnect to it using the debugger.

**Reply by markAtAthena** [reply]
I believe it may happen when it finally tries to hit a breakpoint as I saw some debugging output in my own source get printed and so I moved my breakpoint to the line just before that output and restarted the debugging session and instead of printing that output or breaking, it just disconnected but never printed that debugging output.

**Reply by markAtAthena** [reply]
I'm wondering if I need to incorporate this:

https://github.com/Camelcade/Perl5-IDEA/issues/1771

https://metacpan.org/pod/Apache2::Camelcadedb

https://metacpan.org/release/Plack-Middleware-Camelcadedb

**Reply by hurricup** [reply]
> #38 did get rid of the deep recursion, but now this happens instead as soon as I exercise my code in the same way that triggered the previous deep recursion:
> 
> The debugger disconnects and I see this in my remote ssh terminal where I started the perl process from:
> 
> Buffer has no newlines in it and nothing is in the socket

Debugger tries to read next command from IDE and fails to do so. 
Probabaly there are some errors on IDE side, logs from IDE could be helpful after reproducing. 

Initial problem was that your code has cyclic references (what is bad and may cause a memory leaks) and debugger could not handle this in one place. Probably something else is wrong but I can't say what from the debugger code. 

Ideally would be nice to have an access to the project, but I believe it's not possible :) Or some reproducible sample.

**Reply by hurricup** [reply]
> I'm wondering if I need to incorporate this:
> 
> [Camelcade/Perl5-IDEA#1771](https://github.com/Camelcade/Perl5-IDEA/issues/1771)
> 
> https://metacpan.org/pod/Apache2::Camelcadedb
> 
> https://metacpan.org/release/Plack-Middleware-Camelcadedb

This I can't say. Looks like this makes easier to debug under Apache/Plack. Never used.

**Reply by hurricup** [reply]
I've managed to reproduce the exception :)
